File: sdk/python/guardrail_client.py
import asyncio
import json
import time
from dataclasses import dataclass
from typing import Any, AsyncIterator

import httpx


@dataclass
class GuardrailClient:
    base_url: str
    api_key: str
    timeout: float = 30.0
    with_retry: bool = False
    max_retries: int = 3

    def _payload(
        self,
        prompt: str,
        model: str | None,
        backend: str | None,
        temperature: float,
        max_tokens: int,
    ) -> dict[str, Any]:
        payload = {
            "prompt": prompt,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        if model:
            payload["model"] = model
        if backend:
            payload["backend"] = backend
        return payload

    def _retry_delay(self, response: httpx.Response, attempt: int) -> float:
        retry_after = response.headers.get("Retry-After")
        if retry_after:
            try:
                return float(retry_after)
            except ValueError:
                pass
        return min(2.0 ** (attempt - 1), 8.0)

    def chat(
        self,
        prompt: str,
        *,
        model: str | None = None,
        backend: str | None = None,
        temperature: float = 0.7,
        max_tokens: int = 256,
    ) -> dict[str, Any]:
        url = self.base_url.rstrip("/") + "/chat"
        payload = self._payload(prompt, model, backend, temperature, max_tokens)
        max_retries = self.max_retries if self.with_retry else 0

        for attempt in range(1, max_retries + 2):
            response = httpx.post(
                url,
                headers={"X-Api-Key": self.api_key},
                json=payload,
                timeout=self.timeout,
            )
            if response.status_code in (402, 429) and attempt <= max_retries:
                time.sleep(self._retry_delay(response, attempt))
                continue
            response.raise_for_status()
            return response.json()

    async def chat_stream(
        self,
        prompt: str,
        *,
        model: str | None = None,
        backend: str | None = None,
        temperature: float = 0.7,
        max_tokens: int = 256,
    ) -> AsyncIterator[dict[str, Any]]:
        """
        SSE stream from /chat/stream.

        Yields parsed events — each is a dict with a "type" field:
          {"type":"token",  "content":"..."}
          {"type":"done",   "status":..., "model":..., ...}
          {"type":"blocked","status":"input_blocked"|"output_blocked", ...}
          {"type":"error",  "detail":"..."}
        Retries only happen before the stream starts: once the first chunk
        arrives the connection is never retried.
        """
        url = self.base_url.rstrip("/") + "/chat/stream"
        payload = self._payload(prompt, model, backend, temperature, max_tokens)
        max_retries = self.max_retries if self.with_retry else 0
        attempt = 0

        while True:
            attempt += 1
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                async with client.stream(
                    "POST",
                    url,
                    headers={"X-Api-Key": self.api_key},
                    json=payload,
                ) as response:
                    if response.status_code in (402, 429) and attempt <= max_retries:
                        await asyncio.sleep(self._retry_delay(response, attempt))
                        continue
                    response.raise_for_status()
                    async for line in response.aiter_lines():
                        if line.startswith("data: "):
                            yield json.loads(line[len("data: "):])
                    return

File: sdk/javascript/guardrailClient.mjs
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export class GuardrailClient {
  constructor({ baseUrl, apiKey, withRetry = false, maxRetries = 3 }) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.withRetry = withRetry;
    this.maxRetries = maxRetries;
  }

  _body({ prompt, model, backend, temperature, maxTokens }) {
    return {
      prompt,
      temperature,
      max_tokens: maxTokens,
      ...(model ? { model } : {}),
      ...(backend ? { backend } : {}),
    };
  }

  _retryDelay(retryAfter, attempt) {
    if (retryAfter != null && retryAfter !== "") {
      const seconds = Number(retryAfter);
      if (Number.isFinite(seconds)) {
        return seconds * 1000;
      }
    }
    return Math.min(2 ** (attempt - 1), 8) * 1000;
  }

  async chat({ prompt, model, backend, temperature = 0.7, maxTokens = 256 }) {
    const body = this._body({ prompt, model, backend, temperature, maxTokens });
    const maxRetries = this.withRetry ? this.maxRetries : 0;

    for (let attempt = 1; attempt <= maxRetries + 1; attempt++) {
      const response = await fetch(`${this.baseUrl}/chat`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-Api-Key": this.apiKey,
        },
        body: JSON.stringify(body),
      });

      if ((response.status === 402 || response.status === 429) && attempt <= maxRetries) {
        await sleep(this._retryDelay(response.headers.get("Retry-After"), attempt));
        continue;
      }

      if (!response.ok) {
        throw new Error(await response.text());
      }
      return response.json();
    }
  }

  async *chatStream({ prompt, model, backend, temperature = 0.7, maxTokens = 256 }) {
    /**
     * SSE stream from /chat/stream.
     *
     * Yields parsed events — each is an object with a "type" field:
     *   { type: "token",   content: "..." }
     *   { type: "done",    status: "...", model: "...", ... }
     *   { type: "blocked", status: "input_blocked" | "output_blocked", ... }
     *   { type: "error",   detail: "..." }
     * Retries only happen before the stream starts: once the first chunk
     * arrives the connection is never retried.
     */
    const body = this._body({ prompt, model, backend, temperature, maxTokens });
    const maxRetries = this.withRetry ? this.maxRetries : 0;

    for (let attempt = 1; attempt <= maxRetries + 1; attempt++) {
      const response = await fetch(`${this.baseUrl}/chat/stream`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-Api-Key": this.apiKey,
        },
        body: JSON.stringify(body),
      });

      if ((response.status === 402 || response.status === 429) && attempt <= maxRetries) {
        await sleep(this._retryDelay(response.headers.get("Retry-After"), attempt));
        continue;
      }

      if (!response.ok) {
        throw new Error(await response.text());
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";
      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          break;
        }
        buffer += decoder.decode(value, { stream: true });
        let frameEnd;
        while ((frameEnd = buffer.indexOf("\n\n")) !== -1) {
          const frame = buffer.slice(0, frameEnd);
          buffer = buffer.slice(frameEnd + 2);
          for (const line of frame.split("\n")) {
            if (line.startsWith("data: ")) {
              yield JSON.parse(line.slice("data: ".length));
            }
          }
        }
      }
      return;
    }
  }
}